core/uploader.go

```diff
@@ -134,22 +105,89 @@ func Upload(input io.Reader, outputURI *url.URL, waitBetweenWrites, writeTimeout
 	}
 
 	// We have to do this final write, otherwise there might be final data that's arrived since the last periodic write
-	if _, err := session.SaveData(context.Background(), "", bytes.NewReader(fileContents), fields, writeTimeout); err != nil {
+	if _, _, err := uploadFileWithBackup(outputURI, fileContents, fields, writeTimeout, false, storageFallbackURLs); err != nil {
 		// Don't ignore this error, since there won't be any further attempts to write
 		return nil, fmt.Errorf("failed to write final save: %w", err)
 	}
 	glog.Infof("Completed writing %s to storage", outputURI.Redacted())
 	return nil, nil
 }
 
-func extractThumb(session drivers.OSSession, filename string, segment []byte) error {
+func uploadFileWithBackup(outputURI *url.URL, fileContents []byte, fields *drivers.FileProperties, writeTimeout time.Duration, withRetries bool, storageFallbackURLs map[string]string) (out *drivers.SaveDataOutput, bytesWritten int64, err error) {
+	out, bytesWritten, primaryErr := uploadFile(outputURI, fileContents, fields, writeTimeout, withRetries)
+	if primaryErr == nil {
+		return out, bytesWritten, nil
+	}
+
+	backupURI, err := buildBackupURI(outputURI, storageFallbackURLs)
+	if err != nil {
+		glog.Errorf("failed to build backup URL: %v", err)
+		return nil, 0, primaryErr
+	}
+
+	glog.Warningf("Primary upload failed, uploading to backupURL=%s primaryErr=%q", backupURI.Redacted(), primaryErr)
+	return uploadFile(backupURI, fileContents, fields, writeTimeout, withRetries)
+}
+
+func buildBackupURI(outputURI *url.URL, storageFallbackURLs map[string]string) (*url.URL, error) {
+	outputURIStr := outputURI.String()
+	for primary, backup := range storageFallbackURLs {
+		if strings.HasPrefix(outputURIStr, primary) {
+			backupStr := strings.Replace(outputURIStr, primary, backup, 1)
+			return url.Parse(backupStr)
+		}
+	}
+	return nil, fmt.Errorf("no backup URL found for %s", outputURI.Redacted())
+}
+
+func uploadFile(outputURI *url.URL, fileContents []byte, fields *drivers.FileProperties, writeTimeout time.Duration, withRetries bool) (out *drivers.SaveDataOutput, bytesWritten int64, err error) {
+	outputStr := outputURI.String()
+	// While we wait for storj to implement an easier method for global object deletion we are hacking something
+	// here to allow us to have recording objects deleted after 7 days.
+	if strings.Contains(outputStr, "gateway.storjshare.io/catalyst-recordings-com") {
+		var storjFields drivers.FileProperties
+		if fields != nil {
+			storjFields = *fields
+		}
+		storjFields.Metadata = expiryField
+		fields = &storjFields
```

Comment:
Would this not work:
```suggestion
		if fields == nil {
			fields = &drivers.FileProperties{}
		}
		fields.Metadata = expiryField
```

Reply:
Not really since we would be modifying the `fields` object passed by pointer. So the first attempt (to storj, primary) would change the metadata of the `fields` that are then used for the GS upload. Go can be quite a puzzle sometimes, in TS all I'm doing here could be represented in: `fields = { ...fields, metadata: expiryField }`

Reply:
Ah.. now i see why you named it `storjFields` :)